src/providers/ipa/ipa_hbac_common.c

```diff
@@ -533,56 +536,108 @@ hbac_eval_user_element(TALLOC_CTX *mem_ctx,
+                           attrs,
+                           &m_count, &members);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sysdb_asq_search failed [%d]: %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    if (m_count == 0) {
         DEBUG(SSSDBG_TRACE_LIBS, "No groups for [%s]\n", users->name);
         ret = create_empty_grouplist(users);
         goto done;
     }
-    DEBUG(SSSDBG_TRACE_LIBS,
-          "[%d] groups for [%s]\n", el->num_values, users->name);
+    DEBUG(SSSDBG_TRACE_LIBS, "[%zu] groups for [%s]\n", m_count, username);
 
-    users->groups = talloc_array(users, const char *, el->num_values + 1);
+    users->groups = talloc_array(users, const char *, m_count + 1);
     if (users->groups == NULL) {
         ret = ENOMEM;
         goto done;
     }
 
-    for (i = 0; i < el->num_values; i++) {
-        member_dn = (const char *)el->values[i].data;
+    for (size_t i = 0; i < m_count; i++) {
+        fqgroupname = ldb_msg_find_attr_as_string(members[i], SYSDB_NAME, NULL);
+        if (fqgroupname == NULL) {
+            DEBUG(SSSDBG_MINOR_FAILURE,
+                  "Skipping malformed entry [%s]\n",
+                  ldb_dn_get_linearized(members[i]->dn));
+            continue;
+        }
 
-        ret = get_ipa_groupname(users->groups, domain->sysdb, member_dn,
-                                &users->groups[num_groups]);
-        if (ret != EOK && ret != ERR_UNEXPECTED_ENTRY_TYPE) {
+        /* Only groups from the IPA domain can be referenced from HBAC rules. To
+         * avoid evaluating groups which might even have the same name, but come
+         * from a trusted domain, we first copy the DN to a temporary one..
+         */
+        member_group_container = ldb_dn_copy(tmp_ctx, members[i]->dn);
+        if (member_group_container == NULL) {
+            DEBUG(SSSDBG_MINOR_FAILURE, "ldb_dn_copy failed, skipping!\n");
+            continue;
+        }
+
+        /* ..then remove the RDN which is the group name.. */
+        if (!ldb_dn_remove_child_components(member_group_container, 1)) {
             DEBUG(SSSDBG_MINOR_FAILURE,
-                    "Skipping malformed entry [%s]\n", member_dn);
+                  "ldb_dn_remove_child_components failed, skipping!\n");
             continue;
-        } else if (ret == EOK) {
-            DEBUG(SSSDBG_TRACE_LIBS, "Added group [%s] for user [%s]\n",
-                      users->groups[num_groups], users->name);
-            num_groups++;
+        }
+
+        /* ..the rest should be a DN in the form of:
+         *      cn=groups,cn=ipa_domain_name,cn=sysdb
+         *  If not, just skip this DN. Skipping is safe here, because the rules
+         *  only allow access, never deny, so at worst the user would be denied
+         *  legitimate access
+         */
+        if (ldb_dn_compare(ipa_groups_basedn, member_group_container) != 0) {
+            DEBUG(SSSDBG_FUNC_DATA, "Skipping non-IPA group %s\n", fqgroupname);
```

Comment:
Maybe it would be better to print the DN here as well because it might make it more visible why the groups was skipped. Additionally see comment about fqgroupname below.